library/std/src/sys/sgx/thread.rs

```diff
@@ -3,32 +3,35 @@ use crate::ffi::CStr;
 use crate::io;
 use crate::time::Duration;
 
+use super::abi::tls::ActiveTls;
 use super::abi::usercalls;
 
 pub struct Thread(task_queue::JoinHandle);
 
 pub const DEFAULT_MIN_STACK_SIZE: usize = 4096;
 
 mod task_queue {
-    use crate::sync::mpsc;
+    use super::super::abi::tls::ActiveTls;
+    use super::wait_notify;
     use crate::sync::{Mutex, MutexGuard, Once};
 
-    pub type JoinHandle = mpsc::Receiver<()>;
+    pub type JoinHandle = wait_notify::Waiter;
 
     pub(super) struct Task {
         p: Box<dyn FnOnce()>,
-        done: mpsc::Sender<()>,
+        done: wait_notify::Notifier,
     }
 
     impl Task {
         pub(super) fn new(p: Box<dyn FnOnce()>) -> (Task, JoinHandle) {
-            let (done, recv) = mpsc::channel();
+            let (done, recv) = wait_notify::new();
             (Task { p, done }, recv)
         }
 
-        pub(super) fn run(self) {
+        pub(super) fn run(self, tls_guard: ActiveTls<'_>) {
             (self.p)();
-            let _ = self.done.send(());
+            drop(tls_guard);
+            let _ = self.done.notify();
```

Comment:
I'd prefer moving the notification out of here to entry() (as opposed to moving the TLS destructors in here)

Reply:
Perhaps with some kind of guard that notifies on drop